pinot-common/src/main/java/org/apache/pinot/common/utils/ServiceStatus.java

```diff
@@ -127,34 +129,129 @@ public String getStatusDescription() {
+   * TODO: In this initial version, we are simply adding a configurable static wait time
+   * This can be made smarter:
+   * 1) Keep track of average consumption rate for table in server stats
+   * 2) Monitor consumption rate during startup, report GOOD when it stabilizes to average rate
+   * 3) Monitor consumption rate during startup, report GOOD if it is idle
+   */
+  public static class RealtimeConsumptionCatchupServiceStatusCallback extends BaseServiceStatusCallback {
+
+    private final Map<String, List<String>> _tableToConsumingSegmentsMap;
+    private long _endWaitTime = 0;
+
+    /**
+     * Realtime consumption catchup service which adds a static wait time for
+     */
+    public RealtimeConsumptionCatchupServiceStatusCallback(HelixManager helixManager, String clusterName, String instanceName,
+        long realtimeConsumptionCatchupWaitTimeSeconds) {
+      super(helixManager, clusterName, instanceName);
+
+      _tableToConsumingSegmentsMap = new HashMap<>();
+      for (String resourceName : getResourcesInCluster()) {
+        if (TableNameBuilder.isRealtimeTableResource(resourceName)) {
+
+          IdealState idealState = getResourceIdealState(resourceName);
+          if (idealState.isEnabled()) {
+            List<String> consumingSegments = new ArrayList<>();
+            for (String segmentName : idealState.getPartitionSet()) {
+              Map<String, String> instanceStateMap = idealState.getInstanceStateMap(segmentName);
+              if (CommonConstants.Helix.StateModel.RealtimeSegmentOnlineOfflineStateModel.CONSUMING.equals(
+                  instanceStateMap.get(instanceName))) {
+                consumingSegments.add(segmentName);
+              }
+            }
+            if (!consumingSegments.isEmpty()) {
+              _tableToConsumingSegmentsMap.put(resourceName, consumingSegments);
+            }
+          }
+        }
+      }
+
+      if (_tableToConsumingSegmentsMap.isEmpty()) {
+        LOGGER.info(
+            "No consuming segments to monitor on instance. Setting realtime consumption catchup wait time to 0ms");
+      } else {
+        long startTime = System.currentTimeMillis();
+        _endWaitTime = startTime + TimeUnit.SECONDS.toMillis(realtimeConsumptionCatchupWaitTimeSeconds);
+        LOGGER.info("Monitoring realtime consumption catchup for tables:{}. Will wait for time:{} ms",
+            _tableToConsumingSegmentsMap.keySet(), realtimeConsumptionCatchupWaitTimeSeconds);
+      }
+    }
+
+    @Override
+    public Status getServiceStatus() {
+      long now = System.currentTimeMillis();
+      if (now < _endWaitTime) {
+        _statusDescription = String.format("Waiting for CONSUMING segments to catchup, timeRemaining=%dms", _endWaitTime - System.currentTimeMillis());
+        return Status.STARTING;
+      }
+      _statusDescription = "Wait time for CONSUMING segments catchup expired, now:" + now + ", endWaitTime:" + _endWaitTime;
```

Comment:
I would omit the word 'wait' here (so we can grep for wait and see where waits are happening). How about: "Consuming segments caught up at " + _endWaitTime. 

Or, "Consuming segment status GOOD since " + _endTime.

Reply:
Fixed all suggestions. Please take a look @mcvsubbu